webdriver: Evaluate script commands via the `WebView` API in servoshell

longvatrong111 opened the pull request:
Let `WebDriverCommandMsg::ScriptCommand` goes through embedder first.
Give `embedder` the ability to release `webdriver` from waiting for a response of `ExecuteScript`.

Tests: https://github.com/longvatrong111/servo/actions/runs/16071375821
No regression compared to CI run on main branch.

Fixes: https://github.com/servo/servo/issues/37370

cc: @xiaochengh

longvatrong111 committed:
Move webdriver script command to servoshell

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

wusyong commented:
What's the motivation here. Is there any related issue?

longvatrong111 commented:
> What's the motivation here. Is there any related issue?

This PR is a follow up of https://github.com/servo/servo/pull/36714#issuecomment-2973601211

yezhizhen commented:
I am a bit worried about `forward_webdriver_command`. Commands that used to directly go to Constellation, now goes to embedder which does nothing but forwards it to Constellation. 

I notice that the test execution speed goes down significantly after #36714. There seems to be an extra round of waiting after each subtest. Previously, it almost has no delay. (For example, close_window test which used to took 4 seconds **now took 20 seconds.** And this is a crucial part for every other test.)
cc @mrobinson @jdm @xiaochengh

longvatrong111 commented:
> I am a bit worried about `forward_webdriver_command`. Commands that used to directly go to Constellation, now goes to embedder which does nothing but forwards it to Constellation.
> 
> I notice that the test execution speed goes down significantly after #36714. There seems to be an extra round of waiting after each subtest. Previously, it almost has no delay. (For example, close_window test which used to took 4 seconds **now took 20 seconds.** And this is a crucial part for every other test.) cc @mrobinson @jdm @xiaochengh

We need to move all commands to servoshell then evaluate again.
The reason for `ScriptCommand` was discussed here https://github.com/servo/servo/pull/36714#issuecomment-2973601211.

yezhizhen commented:
I definitely agree that navigation command should be moved.

It is just that, previously it immediately switches to the next subtest.. Right now there is a 1 second gap between each. I'll investigate more once recovered.

> It serves the purpose of synchronization. If script event has shorter route, it may be done out of order with others. In that case, we need to add a wait as well.

That makes sense. Thanks for clarification!

yezhizhen commented:
I definitely agree that navigation command should be moved.

It is just that, previously it immediately switches to the next subtest.. Right now there is a 1 second gap between each. I'll investigate more once recovered.

> It serves the purpose of synchronization. If script event has shorter route, it may be done out of order with others. In that case, we need to add a wait as well.

That makes sense. Thanks for clarification❤️ @longvatrong111

longvatrong111 commented:
> I definitely agree that navigation command should be moved.
> 
> 
> 
> It is just that, previously it immediately switches to the next subtest.. My eyes even cannot catch the speed. Right now there is a 1 second gap between each, leading to greatly increased total time. I'll investigate more once recovered.
> 
> 
> 
> > It serves the purpose of synchronization. If script event has shorter route, it may be done out of order with others. In that case, we need to add a wait as well.
> 
> 
> 
> That makes sense. Thanks for clarification❤️ @longvatrong111 

Let's move this one last. I will also check tests with all pieces of moving commands. Thank you.

mrobinson commented on ports/servoshell/desktop/app.rs:
```diff
@@ -337,10 +337,12 @@ impl App {
                 WebDriverCommandMsg::CloseWebView(webview_id) => {
                     running_state.close_webview(webview_id);
                 },
+                webdriver_msg @ WebDriverCommandMsg::ScriptCommand(..) => {
+                    running_state.forward_webdriver_command(webdriver_msg);
+                },
```
It seems like this should use the exposed `WebView::evaluate_javascript` API and then the WebDriver backend should consume `JSValue` instead of the `WebDriverJSValue`. I know that's a bigger change, but I think it will allow us to remove `WebDriverJSValue` entirely and clean up a lot of the code. Would you be willing to give that a shot?

longvatrong111 replied:
Let me try. Actually we can delay the moving in this PR without any impact.

longvatrong111 commented:
We have more reason to move ScriptCommand to servoshell: https://github.com/servo/servo/issues/35734#issuecomment-3007387495

longvatrong111 commented on ports/servoshell/desktop/app_state.rs:
```diff
@@ -406,6 +408,27 @@ impl RunningAppState {
+            .script_command_interrupt_sender = sender;
+    }
+
+    fn interrupt_webdriver_script_command(&self) {
+        if let Some(sender) = &self
+            .webdriver_senders
+            .borrow()
+            .script_command_interrupt_sender
+        {
+            sender.send(Ok(WebDriverJSValue::Null)).unwrap_or_else(|err| {
+                info!("Notify dialog appear failed. Maybe the channel to webdriver is closed: {err}");
+            });
+        }
```
@mrobinson you mean in the code?
It is quite long too, should be a separate PR:
https://w3c.github.io/webdriver/#user-prompts

longvatrong111 replied:
Ah sorry for the.

> The rules to execute a function body are as follows. The algorithm returns [an ECMAScript completion record](https://w3c.github.io/webdriver/#dfn-completion).
> 
> If at any point during the algorithm a [user prompt](https://w3c.github.io/webdriver/#dfn-user-prompts) appears, immediately return [Completion](https://w3c.github.io/webdriver/#dfn-completion) { [[Type]]: normal, [[Value]]: [null](https://w3c.github.io/webdriver/#dfn-null), [[Target]]: empty }, but continue to run the other steps of this algorithm [in parallel](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel).

It means when an alert box (user prompt) appears, the script `webdriver` sent to `script_thread` continue to run, but `webdriver` immediately returns from the command. In our case, `webdriver` wait for repsonse from `script_thread`, and `embbeder` can help to release `webdriver`.

longvatrong111 committed:
Fix lint

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

longvatrong111 committed:
Update wpt expected result

Signed-off-by: batu_hoang <longvatrong111@gmail.com>

yezhizhen commented:
> Update expected test result based on: https://github.com/longvatrong111/servo/actions/runs/16120879305

This is confusing. Why these tested were added as failed yesterday, but deleted again today?

longvatrong111 commented:
> > Update expected test result based on: https://github.com/longvatrong111/servo/actions/runs/16120879305
> 
> This is confusing. Why these test expectations were newly added yesterday in [9afe027](https://github.com/servo/servo/commit/9afe02756799bf0ce9b092ee175c68b07d563d49), but reverted again today?

I see.
Test result in https://github.com/servo/servo/commit/9afe02756799bf0ce9b092ee175c68b07d563d49 was updated according to local test. After that I ran the CI:
https://github.com/longvatrong111/servo/actions/runs/16089865593

So `/webdriver/tests/classic/perform_actions/key_events.py` and some sub-tests in `perform_actions` were wrongly updated.

Test result in this PR is updated according to https://github.com/longvatrong111/servo/actions/runs/16120879305
This PR actually fixes `element_send_key`, `switch_to_window`. For `perform_actions` it only fixes the wrong update from https://github.com/servo/servo/commit/9afe02756799bf0ce9b092ee175c68b07d563d49

longvatrong111 commented:
Current test result, some unexpected results remain because of intermittent result or CI give different result with local.
https://github.com/longvatrong111/servo/actions/runs/16133927807

longvatrong111 commented:
Hi @mrobinson, I think this one is ready to be merged.

mrobinson commented on ports/servoshell/desktop/app_state.rs:
```diff
@@ -406,6 +408,34 @@ impl RunningAppState {
+    // >
+    // > If at any point during the algorithm a user prompt appears, immediately return
+    // > Completion { Type: normal, Value: null, Target: empty }, but continue to run the
+    // >  other steps of this algorithm in parallel.
```
```suggestion
    /// From <https://w3c.github.io/webdriver/#dfn-execute-a-function-body>:
    /// > The rules to execute a function body are as follows. The algorithm returns
    /// > an ECMAScript completion record.
    // >
    /// > If at any point during the algorithm a user prompt appears, immediately return
    /// > Completion { Type: normal, Value: null, Target: empty }, but continue to run the
    /// >  other steps of this algorithm in parallel.
```

mrobinson committed:
Minor tweaks

- Avoid cloning the entire WebDriver message. Just clone the sender.
- Adjust the documentation to be rustdoc.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>
